### PR TITLE
Aliases pretty-printing and bug fixes

### PR DESCRIPTION
**Pull request by JamesWrigley**
Two changes:
- 4d1c366 adds pretty printing of aliases.
- 7b71869 fixes an embarrassing bug :flushed: 

For the pretty printing, the aliases are grouped by source as much as possible to make the output shorter. Originally I implemented it without newlines in between each entry, but I felt that was bordering on unreadable for proposals with lots of aliases.

<details>
<summary>Example output</summary>

Note: the ✗'s are colored red in a terminal/notebook:
```
Loaded aliases:
  xgm (SA2_XTD1_XGM/XGM/DOOCS):
    xgm-bunches: pulseEnergy.numberOfBunchesActual
    xgm-wavelength: pulseEnergy.wavelengthUsed

  xgm-intensity: SA2_XTD1_XGM/XGM/DOOCS:output[data.intensityTD]

  hirex-pixels: SA2_XTD6_HIREX/CAL/MASTER:daqOutput[data.image.pixels]

  energy-kev: MID_XTD1_UND/DOOCS/ENERGY.actualPosition

  xtd1-transmission: SA2_XTD1_ATT/MDL/MAIN.actual.transmission

  xtd6-transmission: MID_XTD6_ATT/MDL/ATT.actual.transmission

  opt-transmission: MID_AUXT2_ATT/MDL/ATT.actual.transmission

  pp-transmission: MID_LAS_COM/MDL/ATT1_PERCENTS.actualTransmission

  dl800-pos: MID_LAS_COM/MOTOR/DL800.actualPosition

  MID_LAS_COM/CTRL/PPL_PHASE_SHIFTER:
    ppodl-pos: actualPosition
    delay: actualPosition

  MID_EXP_UPP/MDL/DATA_SELECTOR:
    t1-pos: midExpUppMotorT1.actualPosition
    t2-pos: midExpUppMotorT2.actualPosition
    t4-pos: midExpUppMotorT4.actualPosition
    t4-encoderpos: midExpUppMotorT4.encoderPosition
    r8-encoderpos: midExpUppMotorR8.encoderPosition
    sdd: midExpUppMotorT4.encoderPosition
    phi: midExpUppMotorT1.actualPosition
    chi: midExpUppMotorT2.actualPosition
    twotheta: midExpUppMotorR8.encoderPosition

  MID_EXP_SAM/MDL/DATA_SELECTOR:
    sample-x: midExpNafoMotorPpx.actualPosition
    sample-y: midExpNafoMotorPpy.actualPosition
    sample-z: midExpNafoMotorPpz.actualPosition
    ssry-pos: midExpSamMotorSsry.actualPosition
    ssty-pos: midExpSamMotorSsty.actualPosition
    omega: midExpSamMotorSsry.actualPosition

  MID_EXP_SAM/MDL/DATA_SELECTOR_2:
    hexapod-x: hwMidExpSamMotorSshexX.actualPosition
    hexapod-y: hwMidExpSamMotorSshexY.actualPosition
    hexapod-z: hwMidExpSamMotorSshexZ.actualPosition

✗ vacu-MPC: MID_EXP_SAM/GAUGE/PG1.value

  vacu-AGIPD: MID_EXP_AGIPD1M/GAUGE/PG1.value

  lff: MID_EXP_AGIPD1M1/REDU/LITFRM

  lff-output: MID_EXP_AGIPD1M1/REDU/LITFRM:output

  MID_EXP_AGIPD1M1/MDL/DATA_SELECTOR:
    agipd-q1m1-pos: midExpAgipd1MMotorQ1M1.actualPosition
    agipd-q2m1-pos: midExpAgipd1MMotorQ2M1.actualPosition
    agipd-q3m1-pos: midExpAgipd1MMotorQ3M1.actualPosition
    agipd-q4m1-pos: midExpAgipd1MMotorQ4M1.actualPosition

~ agipd0 (MID_DET_AGIPD1M-1/DET/0CH0:xtdf):
    agipd0-data: image.data
  ✗ agipd0-mask: image.mask

~ agipd1 (MID_DET_AGIPD1M-1/DET/1CH0:xtdf):
    agipd1-data: image.data
  ✗ agipd1-mask: image.mask

~ agipd2 (MID_DET_AGIPD1M-1/DET/2CH0:xtdf):
    agipd2-data: image.data
  ✗ agipd2-mask: image.mask

~ agipd3 (MID_DET_AGIPD1M-1/DET/3CH0:xtdf):
    agipd3-data: image.data
  ✗ agipd3-mask: image.mask

~ agipd4 (MID_DET_AGIPD1M-1/DET/4CH0:xtdf):
    agipd4-data: image.data
  ✗ agipd4-mask: image.mask

~ agipd5 (MID_DET_AGIPD1M-1/DET/5CH0:xtdf):
    agipd5-data: image.data
  ✗ agipd5-mask: image.mask

~ agipd6 (MID_DET_AGIPD1M-1/DET/6CH0:xtdf):
    agipd6-data: image.data
  ✗ agipd6-mask: image.mask

~ agipd7 (MID_DET_AGIPD1M-1/DET/7CH0:xtdf):
    agipd7-data: image.data
  ✗ agipd7-mask: image.mask

~ agipd8 (MID_DET_AGIPD1M-1/DET/8CH0:xtdf):
    agipd8-data: image.data
  ✗ agipd8-mask: image.mask

~ agipd9 (MID_DET_AGIPD1M-1/DET/9CH0:xtdf):
    agipd9-data: image.data
  ✗ agipd9-mask: image.mask

~ agipd10 (MID_DET_AGIPD1M-1/DET/10CH0:xtdf):
    agipd10-data: image.data
  ✗ agipd10-mask: image.mask

~ agipd11 (MID_DET_AGIPD1M-1/DET/11CH0:xtdf):
    agipd11-data: image.data
  ✗ agipd11-mask: image.mask

~ agipd12 (MID_DET_AGIPD1M-1/DET/12CH0:xtdf):
    agipd12-data: image.data
  ✗ agipd12-mask: image.mask

~ agipd13 (MID_DET_AGIPD1M-1/DET/13CH0:xtdf):
    agipd13-data: image.data
  ✗ agipd13-mask: image.mask

~ agipd14 (MID_DET_AGIPD1M-1/DET/14CH0:xtdf):
    agipd14-data: image.data
  ✗ agipd14-mask: image.mask

~ agipd15 (MID_DET_AGIPD1M-1/DET/15CH0:xtdf):
    agipd15-data: image.data
  ✗ agipd15-mask: image.mask
```

</details>

Also testable on Maxwell in a definitely-never-used-in-production environment :innocent: 
```bash
# Install Jupyter kernel
$ /gpfs/exfel/u/usr/MID/202304/p003422/Software/analysis_env/bin/python -m ipykernel install --user --name=p3422_analysis --display-name 'p3422 - analysis environment'
```

Proposals with aliases (though they're all mostly the same):
- p3422
- p2798
- p3381
- p3331

### CODE REVIEW

**Comment by tmichela** [comment]
I don't think you should use the "metro" syntax here, since it is nothing we use in extra-data:

```
xgm-intensity: SA2_XTD1_XGM/XGM/DOOCS:output[data.intensityTD]
energy-kev: MID_XTD1_UND/DOOCS/ENERGY.actualPosition
```

I would either use tuple notation:
```
xgm-intensity: (SA2_XTD1_XGM/XGM/DOOCS:output, data.intensityTD)
energy-kev: (MID_XTD1_UND/DOOCS/ENERGY, actualPosition)
```

or same as with multiple alisases:
```
SA2_XTD1_XGM/XGM/DOOCS:output
  xgm-intensity: data.intensityTD
MID_XTD1_UND/DOOCS/ENERGY
  energy-kev: actualPosition
```

**Comment by JamesWrigley** [comment]
Makes sense, switched to using tuples in 286acbf. I originally used the same format as with multiple aliases, but that made it look even more cluttered so I went for a one-liner instead.

File: extra_data/reader.py [hunk 1]
```diff
@@ -1633,6 +1633,107 @@ def __contains__(self, aliased_item):
...
 
+    def __repr__(self):
+        """
+        Pretty-print all the aliases.
+        """
+        RED = "\033[91m"
+        END_COLOR = "\033[0m"
+
+        # Check if the alias exists
+        def alias_exists(alias):
```
Comment (philsmt) [inline extra_data/reader.py:1644]: Why not use `AliasIndexer.__contains__` added in https://github.com/European-XFEL/EXtra-data/pull/398?

File: extra_data/reader.py [hunk 2]
```diff
@@ -1633,6 +1633,107 @@ def __contains__(self, aliased_item):
...
+
+        # Get the right icon for an alias
+        def alias_icon(exists):
+            if isinstance(exists, str):
+                exists = alias_exists(exists)
+
+            return " " if exists else f"{RED}✗{END_COLOR}"
+
+        # Find the alias for a source, if one exists
+        def source_alias(source):
```
Comment (philsmt) [inline extra_data/reader.py:1659]: I wonder if such a method may actually be public, `run.alias.search('<source>')`
Ultimately this PR also offers a way to do this, but at some point large alias definition may becomes unwieldy that people want to filter interactively in a notebook?

**Review by philsmt — COMMENTED** [review]
Thanks, looks good apart from the minor comment regarding `alias_exists`.

We're really reaching a point where `extra_data.reader` becomes absurdly large. While I don't think it's enough to fix the problem, maybe `AliasIndexer` is becoming large enough to warrant its own module?

**Comment by takluyver** [comment]
I largely second what @philsmt said, though I wouldn't make a public `.search()` method at this point. Not to rule it out, just I don't think the need is clear yet, and if we do do it, there's a question over what you might specify to find aliases (e.g. source name, key name, substrings, glob patterns, regexes...).

**Comment by JamesWrigley** [comment]
Moved `AliasIndexer` to a new file in c7bc558.

**Pull request merged**